File: logitech_cam/src/command_generator.cc
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "CreeperCam.h"
#include "logitech_cam/CamCmd.h"
#include <iostream>
using namespace std;

/*
 *
 * Adam Nilsson, 7/2/12
 */


int main(int argc, char **argv)
{

	ROS_INFO("Random Command Generator Starting");
	//ROS_INFO(argv[1]);
	ros::init(argc, argv, "command_generator");

	ros::NodeHandle n;
	ros::Publisher chatter_pub = n.advertise<logitech_cam::CamCmd>("/logitech_cam/camera_instr", 1000);
	ros::Rate loop_rate(1);
	int count = 0;

	logitech_cam::CamCmd commands[6];
	commands[0].Pvalue = 250;
	commands[0].Tvalue = 0;
	commands[0].Zvalue = 0;
	commands[1].Pvalue = -250;
	commands[1].Tvalue = 0;
	commands[1].Zvalue = 0;
	commands[2].Pvalue = 0;
	commands[2].Tvalue = 150;
	commands[2].Zvalue = 0;
	commands[3].Pvalue = 0;
	commands[3].Tvalue = -150;
	commands[3].Zvalue = 0;
	commands[4].Pvalue = 0;
	commands[4].Tvalue = 0;
	commands[4].Zvalue = -10;
	commands[5].Pvalue = 0;
	commands[5].Tvalue = 0;
	commands[5].Zvalue = 10;

//	commands[4].Pvalue = 250;
//	commands[4].Tvalue = 0;
//	commands[4].Zvalue = 0;
//	commands[5].Pvalue = -250;
//	commands[5].Tvalue = 0;
//	commands[5].Zvalue = 0;
//	commands[6].Pvalue = 0;
//	commands[6].Tvalue = 150;
//	commands[6].Zvalue = 0;
//	commands[7].Pvalue = 0;
//	commands[7].Tvalue = -150;
//	commands[7].Zvalue = 0;
//
//	commands[8].Pvalue = 250;
//	commands[8].Tvalue = 0;
//	commands[8].Zvalue = 10;
//	commands[9].Pvalue = -250;
//	commands[9].Tvalue = 0;
//	commands[9].Zvalue = 10;
//	commands[10].Pvalue = 0;
//	commands[10].Tvalue = 150;
//	commands[10].Zvalue = 10;
//	commands[11].Pvalue = 0;
//	commands[11].Tvalue = -150;
//	commands[11].Zvalue = 10;

	ros::Duration(10).sleep();
	ROS_INFO("Ready to send commands");
	srand(time(NULL));
	while (ros::ok())
	{
		ROS_INFO("going to send message");
		int ci = rand() % 6;
		chatter_pub.publish(commands[ci]);
		ros::spinOnce();

		loop_rate.sleep();
	}

	return 0;
}

File: logitech_cam/src/command_generator.py
#!/usr/bin/env python
import roslib 
import sys
roslib.load_manifest('logitech_cam')
import rospy
import random
from std_msgs.msg import String
from logitech_cam.msg import IntArray


def parse_commands(str):
	cmd_strs = str.strip('[]').split('],[')
	num_cmd = len(cmd_strs)
	
	cmd_list = [[]]*num_cmd
	
	for i in range(num_cmd):
		cmd_list[i] = map(int,cmd_strs[i].split(','))
	
	print 'Command List Parsed:'
	print '============================='
	print 'Found ',num_cmd,' commands:'
	print cmd_list
	return cmd_list

# Generates a IntArray message with the command data given in intarray
def get_msg(intarray):
	msg = IntArray()
	msg.data = intarray
	return msg
	

def main(args):
	rospy.init_node('command_generator', anonymous=True)
	#print args
	# Test string for commands
#	if rospy.has_param('/command_generator/logitech_cam/command_list'):
#		cmd_str = rospy.get_param('/command_generator/command_list')
#	else:
	# Check for arguments
	try:
		cmd_str = args[args.index('-c')+1]
	except ValueError:
		print 'No given commands found, using default command string' 
		cmd_str = '[[150,0,0],[-150,0,0],[0,100,0],[0,-100,0],[0,0,10],[0,0,-10]]'
	
	try:
		duration = int(args[args.index('-t')+1])
		print 'Running for ',duration,'s'
	except ValueError:
		print 'No duration found, running for 60s.'
		duration = 60
	
	# Generation frequency
	try:
		freq = float(args[args.index('-f')+1])
	except ValueError:
		freq = 1.0
	print freq
	
	
	# Start at random position ?
	try:
		start_random = args[args.index('-r')+1].lower() in ("yes", "true", "t", "1")
	except ValueError:
		start_random = False
	print 'start random position: 	',start_random
	
	time0 = rospy.get_time()
	
	cmd_list = parse_commands(cmd_str)
	num_cmd = len(cmd_list)
	cmd_publisher = rospy.Publisher('/logitech_cam/camera_instr',IntArray)
	
	# Initiate random number generator  
	random.seed()
	
	# Set the frequence for sending commands
	### make rate as parameter
	r = rospy.Rate(freq)
	while not rospy.is_shutdown():
		while rospy.get_time()-time0 < duration:
			ri = random.randint(0,num_cmd-1)
			print rospy.get_time()-time0
			this_cmd = cmd_list[ri]
			print 'Sending command: 	',this_cmd
			cmd_publisher.publish(get_msg(this_cmd))
			r.sleep()
		
	print 'Shutting down'

if __name__ == '__main__':
	main(sys.argv)
